packages/firestore/src/core/sync_engine.ts

```diff
@@ -801,6 +834,12 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
     state: QueryTargetState,
     error?: FirestoreError
   ): Promise<void> {
+    if (this.isPrimary) {
+      // If we receive a target state notification via Web Storage, we are
+      // either already secondary or another tab has taken the primary lease.
+      await this.applyPrimaryState(false);
```

Comment:
Removing these lines breaks some of the new tests. This is certainly a test issue, but in general multi-tab is only safe if there is only one primary (we don't want two simultaneous WriteStreams for example). I prefer having no primary temporarily than two primaries at once.

Reply:
I'd prefer that too, but I don't think we can guarantee it.  The only thing we can guarantee is that only one tab can make IndexedDb writes as primary at a time (via transactions that require the primary lease).  But we _could_ have two tabs with WriteStreams open for some period of time.

Ideally our design would allow for this.  If it's infeasible, then I guess we should acknowledge that there are race conditions but convince ourselves that it'll be sufficiently unlikely and/or non-catastrophic or something...